### PR TITLE
Changed the train! function to support the syntax of the new explicit format. 

### PR DESCRIPTION
**Pull request by Roeya**
Hi Lior
I have changed the syntax to the new explicit format of Flux 0.14.
I also changed the demo.jl script to use path in none OS dependency (joinpath) and relative to its location using @__DIR__

### CODE REVIEW

File: examples/demo.jl [hunk 1]
```diff
@@ -15,7 +14,7 @@ using TransformersLite
 include("utilities.jl")
 include("training.jl")
 
-path = "datasets\\amazon_reviews_multi\\en\\1.0.0\\"
+path = joinpath(@__DIR__, "..", "datasets", "amazon_reviews_multi", "en", "1.0.0")
```
Comment (LiorSinai) [inline examples/demo.jl]: The readme currently says change to this directory first, then the ".." won't be needed. So either this should be updated or the readme.

File: examples/training.jl [hunk 2]
```diff
@@ -48,34 +48,34 @@ function split_validation(X::AbstractVector, Y::AbstractVector, frac=0.1; rng=Ra
     (X[idxs[1:ntrain]], Y[idxs[1:ntrain]]), (X[idxs[ntrain+1:end]], Y[idxs[ntrain+1:end]])
 end
 
-function train!(loss, ps, data, opt, val_data; n_epochs=10)
+function train!(loss, model, train_data, opt, val_data; n_epochs=10)
```
Comment (LiorSinai) [inline examples/training.jl]: Think opt should be called opt_state instead

File: examples/training.jl [hunk 3]
```diff
@@ -48,34 +48,34 @@ function split_validation(X::AbstractVector, Y::AbstractVector, frac=0.1; rng=Ra
...
         "val_loss"=>Float64[]
         )
     for epoch in 1:n_epochs
-        ps = Flux.Params(ps)
-        progress = Progress(length(data); desc="epoch $epoch/$n_epochs")
-        for Xy in data
-            gs = gradient(ps) do
-                loss(Xy)
+        progress = Progress(length(train_data); desc="epoch $epoch/$n_epochs")
+        epoch_loss = 0.0
```
Comment (LiorSinai) [inline examples/training.jl]: This epoch_loss is just left hanging here. But it might be worthwhile calculating it as the mean or sum of the batch losses.

To get the batch loss `val` do:
```
val, grads = Flux.withgradient(model) do m
    loss(m(Xy[1]), Xy[2])
end
```

**Review by LiorSinai — CHANGES_REQUESTED** [review]
Overall looks good, just a few small comments.

Could you also add a compat in the project.toml for Flux > 0.13.9 since the main goal is to use new Flux syntax. Think that is also enough of a breaking change to justify changing the version to 0.2.0

File: examples/demo.jl [hunk 4]
```diff
@@ -120,31 +119,31 @@ hyperparameters["model"] = "$(typeof(model).name.wrapper)"
...
+    loss(x, y) = Flux.logitcrossentropy(x, y)
     accuracy(ŷ, y) = mean(Flux.onecold(ŷ) .== Flux.onecold(y))
 end
 
-loss(x::Tuple) = loss(x[1], x[2])
+
 
 ## Training 
-opt = ADAM()
+opt = Flux.setup(Flux.Adam(), model)
```
Comment (LiorSinai) [inline examples/demo.jl]: `opt` should be called `opt_state`

File: examples/training.jl (hunk 2 above)
Comment (Roeya) [reply examples/training.jl]: done

File: examples/training.jl (hunk 3 above)
Comment (Roeya) [reply examples/training.jl]: done, added more loss info during progress reporting

File: examples/demo.jl (hunk 4 above)
Comment (Roeya) [reply examples/demo.jl]: done

**Review by LiorSinai — APPROVED** [review]

File: examples/demo.jl [hunk 5]
```diff
@@ -120,31 +121,29 @@ hyperparameters["model"] = "$(typeof(model).name.wrapper)"
...
+    loss(x, y) = Flux.logitcrossentropy(x, y)
     accuracy(ŷ, y) = mean(Flux.onecold(ŷ) .== Flux.onecold(y))
 end
 
-loss(x::Tuple) = loss(x[1], x[2])
-
-## Training 
-opt = ADAM()
+## Training (we use Optimisers instead of Flux due to bug in Flux below v0.13.11)
+opt_state = Optimisers.setup(Optimisers.Adam(), model)
```
Comment (LiorSinai) [inline examples/demo.jl:132]: The workaround is good but small question: why not target 0.13.11 instead?

My original idea was for 0.13.9 because that would allow the new syntax while having the broadest compatibility. Version 0.13.11 is only 2 patch versions ahead, so it won't make a big difference.

Version 0.14 is a different story. It has several breaking changes and requires the new Julia 1.9. A few of these are with package management - I see you encountered those errors in another commit. So I am hesitant to upgrade to it just yet.

Comment (Roeya) [reply examples/demo.jl:132]: Just followed your request to be compatible with 0.13.9 as I thought there is some reason for this specific version.
